OIDC client: quarkus.oidc-client.enabled does not working in combinatie with @RegisterProvider(OidcClientRequestReactiveFilter.class).

### Describe the bug

While testing you can disabled the oidc client by setting quarkus.oidc-client.enabled to false. This work perfectly in combination with @OidcClientFilter. But not when you are using @RegisterProvider using the default class or custom class. 



### Expected behavior

oidc Client is disabled when using @RegisterProvider

### Actual behavior

	Caused by: java.lang.NullPointerException: Cannot invoke "io.quarkus.oidc.client.OidcClients.getClient()" because "oidcClients" is null
		at io.quarkus.oidc.client.runtime.AbstractTokensProducer.init(AbstractTokensProducer.java:43)
		at io.quarkus.oidc.client.reactive.filter.OidcClientRequestReactiveFilter_Bean.doCreate(Unknown Source)
		at io.quarkus.oidc.client.reactive.filter.OidcClientRequestReactiveFilter_Bean.create(Unknown Source)
		at io.quarkus.oidc.client.reactive.filter.OidcClientRequestReactiveFilter_Bean.get(Unknown Source)
		at io.quarkus.oidc.client.reactive.filter.OidcClientRequestReactiveFilter_Bean.get(Unknown Source)
		at io.quarkus.arc.impl.ArcContainerImpl.beanInstanceHandle(ArcContainerImpl.java:559)
		at io.quarkus.arc.impl.ArcContainerImpl.beanInstanceHandle(ArcContainerImpl.java:539)
		at io.quarkus.arc.impl.ArcContainerImpl.beanInstanceHandle(ArcContainerImpl.java:572)
		at io.quarkus.arc.impl.ArcContainerImpl.instanceHandle(ArcContainerImpl.java:534)
		at io.quarkus.arc.impl.ArcContainerImpl.instance(ArcContainerImpl.java:294)
		at io.quarkus.rest.client.reactive.runtime.RestClientBuilderImpl.register(RestClientBuilderImpl.java:292)
		at io.quarkus.rest.client.reactive.runtime.RestClientBuilderImpl.build(RestClientBuilderImpl.java:408)
		at io.quarkus.rest.client.reactive.runtime.QuarkusRestClientBuilderImpl.build(QuarkusRestClientBuilderImpl.java:267)
		at io.quarkus.rest.client.reactive.runtime.RestClientCDIDelegateBuilder.build(RestClientCDIDelegateBuilder.java:67)
		at io.quarkus.rest.client.reactive.runtime.RestClientCDIDelegateBuilder.createDelegate(RestClientCDIDelegateBuilder.java:49)
		at io.quarkus.rest.client.reactive.runtime.RestClientReactiveCDIWrapperBase.delegate(RestClientReactiveCDIWrapperBase.java:76)
		at io.quarkus.rest.client.reactive.runtime.RestClientReactiveCDIWrapperBase.<init>(RestClientReactiveCDIWrapperBase.java:30)

### How to Reproduce?

1. create oidc client and disabled it
2. activate the client by using the annotation @RegisterProvider

### Output of `uname -a` or `ver`

Microsoft Windows [Version 10.0.19045.4529]

### Output of `java -version`

openjdk 21.0.2 2024-01-16 LTS OpenJDK Runtime Environment Corretto-21.0.2.13.1 (build 21.0.2+13-LTS) OpenJDK 64-Bit Server VM Corretto-21.0.2.13.1 (build 21.0.2+13-LTS, mixed mode, sharing)

### Quarkus version or git rev

3.12.2

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

3.9.6

### Additional information

_No response_

--- a/extensions/oidc-client-filter/runtime/src/main/java/io/quarkus/oidc/client/filter/runtime/AbstractOidcClientRequestFilter.java
+++ b/extensions/oidc-client-filter/runtime/src/main/java/io/quarkus/oidc/client/filter/runtime/AbstractOidcClientRequestFilter.java
@@ -21,6 +21,11 @@ public class AbstractOidcClientRequestFilter extends AbstractTokensProducer impl
 
     @Override
     public void filter(ClientRequestContext requestContext) throws IOException {
+        if (isClientFeatureDisabled()) {
+            LOG.debug("OIDC client filter can not acquire and propagate tokens because "
+                    + "OIDC client is disabled with `quarkus.oidc-client.enabled=false`");
+            return;
+        }
         try {
             final String accessToken = getAccessToken();
             requestContext.getHeaders().add(HttpHeaders.AUTHORIZATION, BEARER_SCHEME_WITH_SPACE + accessToken);

--- a/extensions/oidc-client-reactive-filter/deployment/src/test/java/io/quarkus/oidc/client/reactive/filter/ClientWebApplicationExceptionMapper.java
+++ b/extensions/oidc-client-reactive-filter/deployment/src/test/java/io/quarkus/oidc/client/reactive/filter/ClientWebApplicationExceptionMapper.java
@@ -1,0 +1,17 @@
+package io.quarkus.oidc.client.reactive.filter;
+
+import jakarta.ws.rs.core.Response;
+import jakarta.ws.rs.ext.ExceptionMapper;
+import jakarta.ws.rs.ext.Provider;
+
+import org.jboss.resteasy.reactive.ClientWebApplicationException;
+
+@Provider
+public class ClientWebApplicationExceptionMapper implements ExceptionMapper<ClientWebApplicationException> {
+
+    @Override
+    public Response toResponse(ClientWebApplicationException t) {
+        return Response.status(t.getResponse().getStatus()).build();
+    }
+
+}

--- a/extensions/oidc-client-reactive-filter/deployment/src/test/java/io/quarkus/oidc/client/reactive/filter/NamedOidcClientFilterDevModeTest.java
+++ b/extensions/oidc-client-reactive-filter/deployment/src/test/java/io/quarkus/oidc/client/reactive/filter/NamedOidcClientFilterDevModeTest.java
@@ -18,7 +18,8 @@ public class NamedOidcClientFilterDevModeTest {
             ProtectedResourceServiceAnnotationOidcClient.class,
             ProtectedResourceServiceConfigPropertyOidcClient.class,
             ProtectedResourceServiceCustomProviderConfigPropOidcClient.class,
-            OidcClientResource.class
+            OidcClientResource.class,
+            ClientWebApplicationExceptionMapper.class
     };
 
     @RegisterExtension
@@ -31,8 +32,47 @@ public class NamedOidcClientFilterDevModeTest {
     public void testGerUserConfigPropertyAndAnnotation() {
         // test OidcClientFilter with OidcClient selected via annotation or config-property
 
+        // Client feature is disabled
         // OidcClient selected via @OidcClient("clientName")
         RestAssured.when().get("/oidc-client/annotation/user-name")
+                .then()
+                .statusCode(401);
+
+        RestAssured.when().get("/oidc-client/annotation/anonymous-user-name")
+                .then()
+                .statusCode(204)
+                .body(equalTo(""));
+
+        // @OidcClientFilter: OidcClient selected via `quarkus.oidc-client-filter.client-name=config-property`
+        RestAssured.when().get("/oidc-client/config-property/user-name")
+                .then()
+                .statusCode(401);
+
+        RestAssured.when().get("/oidc-client/config-property/anonymous-user-name")
+                .then()
+                .statusCode(204)
+                .body(equalTo(""));
+
+        // @RegisterProvider(OidcClientRequestReactiveFilter.class): OidcClient selected via `quarkus.oidc-client-filter.client-name=config-property`
+        RestAssured.when().get("/oidc-client/custom-provider-config-property/user-name")
+                .then()
+                .statusCode(401);
+
+        RestAssured.when().get("/oidc-client/custom-provider-config-property/anonymous-user-name")
+                .then()
+                .statusCode(204)
+                .body(equalTo(""));
+
+        test.modifyResourceFile("application.properties", s -> s.replace(".enabled=false", ".enabled=true"));
+
+        // Client feature is enabled
+        // OidcClient selected via @OidcClient("clientName")
+        RestAssured.when().get("/oidc-client/annotation/user-name")
+                .then()
+                .statusCode(200)
+                .body(equalTo("jdoe"));
+
+        RestAssured.when().get("/oidc-client/annotation/anonymous-user-name")
                 .then()
                 .statusCode(200)
                 .body(equalTo("jdoe"));
@@ -43,8 +83,17 @@ public class NamedOidcClientFilterDevModeTest {
                 .statusCode(200)
                 .body(equalTo("alice"));
 
+        RestAssured.when().get("/oidc-client/config-property/anonymous-user-name")
+                .then()
+                .statusCode(200)
+                .body(equalTo("alice"));
+
         // @RegisterProvider(OidcClientRequestReactiveFilter.class): OidcClient selected via `quarkus.oidc-client-filter.client-name=config-property`
         RestAssured.when().get("/oidc-client/custom-provider-config-property/user-name")
+                .then()
+                .statusCode(200)
+                .body(equalTo("alice"));
+        RestAssured.when().get("/oidc-client/custom-provider-config-property/anonymous-user-name")
                 .then()
                 .statusCode(200)
                 .body(equalTo("alice"));

--- a/extensions/oidc-client-reactive-filter/deployment/src/test/java/io/quarkus/oidc/client/reactive/filter/OidcClientResource.java
+++ b/extensions/oidc-client-reactive-filter/deployment/src/test/java/io/quarkus/oidc/client/reactive/filter/OidcClientResource.java
@@ -28,14 +28,32 @@ public class OidcClientResource {
     }
 
     @GET
+    @Path("/annotation/anonymous-user-name")
+    public String annotationAnonymousUserName() {
+        return protectedResourceServiceAnnotationOidcClient.getAnonymousUserName();
+    }
+
+    @GET
     @Path("/config-property/user-name")
     public String configPropertyUserName() {
         return protectedResourceServiceConfigPropertyOidcClient.getUserName();
     }
 
     @GET
+    @Path("/config-property/anonymous-user-name")
+    public String configPropertyAnonymousUserName() {
+        return protectedResourceServiceConfigPropertyOidcClient.getAnonymousUserName();
+    }
+
+    @GET
     @Path("/custom-provider-config-property/user-name")
     public String customProviderConfigPropertyUserName() {
         return protectedResourceServiceCustomProviderConfigPropOidcClient.getUserName();
+    }
+
+    @GET
+    @Path("/custom-provider-config-property/anonymous-user-name")
+    public String customProviderConfigPropertyAnonymousUserName() {
+        return protectedResourceServiceCustomProviderConfigPropOidcClient.getAnonymousUserName();
     }
 }

--- a/extensions/oidc-client-reactive-filter/deployment/src/test/java/io/quarkus/oidc/client/reactive/filter/ProtectedResource.java
+++ b/extensions/oidc-client-reactive-filter/deployment/src/test/java/io/quarkus/oidc/client/reactive/filter/ProtectedResource.java
@@ -7,10 +7,7 @@ import jakarta.inject.Inject;
 import jakarta.ws.rs.GET;
 import jakarta.ws.rs.Path;
 
-import io.quarkus.security.Authenticated;
-
 @Path("/protected")
-@Authenticated
 public class ProtectedResource {
 
     @Inject
@@ -19,6 +16,12 @@ public class ProtectedResource {
     @GET
     @RolesAllowed("user")
     public String principalName() {
+        return principal.getName();
+    }
+
+    @GET
+    @Path("/anonymous")
+    public String anonymousPrincipalName() {
         return principal.getName();
     }
 }

--- a/extensions/oidc-client-reactive-filter/deployment/src/test/java/io/quarkus/oidc/client/reactive/filter/ProtectedResourceServiceAnnotationOidcClient.java
+++ b/extensions/oidc-client-reactive-filter/deployment/src/test/java/io/quarkus/oidc/client/reactive/filter/ProtectedResourceServiceAnnotationOidcClient.java
@@ -14,4 +14,8 @@ public interface ProtectedResourceServiceAnnotationOidcClient {
 
     @GET
     String getUserName();
+
+    @GET
+    @Path("/anonymous")
+    String getAnonymousUserName();
 }

--- a/extensions/oidc-client-reactive-filter/deployment/src/test/java/io/quarkus/oidc/client/reactive/filter/ProtectedResourceServiceConfigPropertyOidcClient.java
+++ b/extensions/oidc-client-reactive-filter/deployment/src/test/java/io/quarkus/oidc/client/reactive/filter/ProtectedResourceServiceConfigPropertyOidcClient.java
@@ -14,4 +14,8 @@ public interface ProtectedResourceServiceConfigPropertyOidcClient {
 
     @GET
     String getUserName();
+
+    @GET
+    @Path("/anonymous")
+    String getAnonymousUserName();
 }

--- a/extensions/oidc-client-reactive-filter/deployment/src/test/java/io/quarkus/oidc/client/reactive/filter/ProtectedResourceServiceCustomProviderConfigPropOidcClient.java
+++ b/extensions/oidc-client-reactive-filter/deployment/src/test/java/io/quarkus/oidc/client/reactive/filter/ProtectedResourceServiceCustomProviderConfigPropOidcClient.java
@@ -13,4 +13,8 @@ public interface ProtectedResourceServiceCustomProviderConfigPropOidcClient {
 
     @GET
     String getUserName();
+
+    @GET
+    @Path("/anonymous")
+    String getAnonymousUserName();
 }

--- a/extensions/oidc-client-reactive-filter/deployment/src/test/resources/application-oidc-client-reactive-filter.properties
+++ b/extensions/oidc-client-reactive-filter/deployment/src/test/resources/application-oidc-client-reactive-filter.properties
@@ -2,6 +2,8 @@ quarkus.oidc.auth-server-url=${keycloak.url}/realms/quarkus/
 quarkus.oidc.client-id=quarkus-service-app
 quarkus.oidc.credentials.secret=secret
 
+quarkus.oidc-client.enabled=false
+
 quarkus.rest-client-oidc-filter.client-name=config-property
 quarkus.oidc-client.config-property.auth-server-url=${quarkus.oidc.auth-server-url}
 quarkus.oidc-client.config-property.client-id=${quarkus.oidc.client-id}

--- a/extensions/oidc-client-reactive-filter/runtime/src/main/java/io/quarkus/oidc/client/reactive/filter/runtime/AbstractOidcClientRequestReactiveFilter.java
+++ b/extensions/oidc-client-reactive-filter/runtime/src/main/java/io/quarkus/oidc/client/reactive/filter/runtime/AbstractOidcClientRequestReactiveFilter.java
@@ -26,6 +26,11 @@ public class AbstractOidcClientRequestReactiveFilter extends AbstractTokensProdu
 
     @Override
     public void filter(ResteasyReactiveClientRequestContext requestContext) {
+        if (isClientFeatureDisabled()) {
+            LOG.debug("OIDC client filter can not acquire and propagate tokens because "
+                    + "OIDC client is disabled with `quarkus.oidc-client.enabled=false`");
+            return;
+        }
         requestContext.suspend();
 
         super.getTokens().subscribe().with(new Consumer<>() {

--- a/extensions/oidc-client/runtime/src/main/java/io/quarkus/oidc/client/runtime/AbstractTokensProducer.java
+++ b/extensions/oidc-client/runtime/src/main/java/io/quarkus/oidc/client/runtime/AbstractTokensProducer.java
@@ -24,11 +24,18 @@ public abstract class AbstractTokensProducer {
 
     @Inject
     public OidcClientsConfig oidcClientsConfig;
+    @Inject
+    public OidcClientBuildTimeConfig oidcClientBuildTimeConfig;
 
     final TokensHelper tokensHelper = new TokensHelper();
 
     @PostConstruct
     public void init() {
+        if (isClientFeatureDisabled()) {
+            LOG.debug("OIDC client is disabled with `quarkus.oidc-client.enabled=false`,"
+                    + " skipping the token producer initialization");
+            return;
+        }
         Optional<OidcClient> initializedClient = client();
         if (initializedClient.isEmpty()) {
             Optional<String> clientId = Objects.requireNonNull(clientId(), "clientId must not be null");
@@ -49,13 +56,25 @@ public abstract class AbstractTokensProducer {
         initTokens();
     }
 
+    protected boolean isClientFeatureDisabled() {
+        return !oidcClientBuildTimeConfig.enabled;
+    }
+
     protected void initTokens() {
+        if (isClientFeatureDisabled()) {
+            throw new IllegalStateException("OIDC client feature is disabled with `quarkus.oidc-client.enabled=false`"
+                    + " but the initTokens() method is called.");
+        }
         if (earlyTokenAcquisition) {
             tokensHelper.initTokens(oidcClient, additionalParameters());
         }
     }
 
     public Uni<Tokens> getTokens() {
+        if (isClientFeatureDisabled()) {
+            throw new IllegalStateException("OIDC client feature is disabled with `quarkus.oidc-client.enabled=false`"
+                    + " but the getTokens() method is called.");
+        }
         final boolean forceNewTokens = isForceNewTokens();
         if (forceNewTokens) {
             final Optional<String> clientId = clientId();
@@ -66,6 +85,9 @@ public abstract class AbstractTokensProducer {
     }
 
     public Tokens awaitTokens() {
+        if (isClientFeatureDisabled()) {
+            throw new IllegalStateException("OIDC client feature is disabled with `quarkus.oidc-client.enabled=false`.");
+        }
         return getTokens().await().indefinitely();
     }
 

--- a/extensions/oidc-client/runtime/src/main/java/io/quarkus/oidc/client/runtime/OidcClientBuildTimeConfig.java
+++ b/extensions/oidc-client/runtime/src/main/java/io/quarkus/oidc/client/runtime/OidcClientBuildTimeConfig.java
@@ -1,12 +1,13 @@
 package io.quarkus.oidc.client.runtime;
 
 import io.quarkus.runtime.annotations.ConfigItem;
+import io.quarkus.runtime.annotations.ConfigPhase;
 import io.quarkus.runtime.annotations.ConfigRoot;
 
 /**
  * Build time configuration for OIDC client.
  */
-@ConfigRoot
+@ConfigRoot(name = "oidc-client", phase = ConfigPhase.BUILD_AND_RUN_TIME_FIXED)
 public class OidcClientBuildTimeConfig {
     /**
      * If the OIDC client extension is enabled.

